src/controllers/ComputedValues/ComputedValue.js

```diff
@@ -0,0 +1,140 @@
+	ERROR: 'error'
+});
+
+function defaultShouldCompute(prevDependencies, currDependencies) {
+	// If no dependencies exist yet, then this is the first update
+	if (prevDependencies === null) {
+		return true;
+	}
+
+	return prevDependencies.some((prev, index) => prev !== currDependencies[index]);
+}
+
+export default class ComputedValue {
+	computeComplete = Promise.resolve();
+	error = null;
+	host;
+	pending = false;
+	success = null;
+	value;
+
+	constructor(host, options = {}) {
+		const {
+			compute,
+			shouldCompute = defaultShouldCompute,
+			getDependencies,
+			isAsync = false,
+			initialValue
+		} = options;
+
+		if (typeof compute !== 'function') {
+			console.error('Failed to initialize ComputedValueController: compute parameter must be a function.');
+			return;
+		}
+
+		if (typeof getDependencies !== 'function') {
+			console.error('Failed to initialize ComputedValueController: getDependencies parameter must be a function.');
+			return;
+		}
+
+		this._compute = compute;
+		this._shouldCompute = shouldCompute;
+		this._getDependencies = getDependencies;
+		this._isAsync = isAsync;
+		this.value = initialValue;
+
+		(this.host = host).addController(this);
+	}
+
+	get asyncStatus() {
+		if (this.pending) {
+			return ASYNC_STATUSES.PENDING;
+		}
+
+		if (this.success === null) {
+			return ASYNC_STATUSES.INITIAL;
+		}
+
+		if (this.success) {
+			return ASYNC_STATUSES.SUCCESS;
+		}
+
+		return ASYNC_STATUSES.ERROR;
+	}
+
+	hostUpdate() {
+		const currDependencies = this._getDependencies();
+
+		if (this._shouldCompute(this._prevDependencies, currDependencies)) {
+			this._prevDependencies = currDependencies;
+			this._updateValue(currDependencies);
+		}
+	}
+
+	_compute;
+	_computeCompleteResolve = () => {};
+	_getDependencies;
+	_isAsync;
+	_latestPromise;
+	_prevDependencies = null;
+	_shouldCompute;
+
+	_updateAsyncValue(dependencies) {
+		this._latestPromise = null; // Prevent any ongoing compute update from completing as this is the latest now
+
+		// If there's still an ongoing compute update, don't replace the computeComplete promise.
+		// Instead, this effectively extends the computeComplete promise resolution until the _latestPromise is done.
+		if (this.pending === false) {
+			this.computeComplete = new Promise((resolve) => {
+				this._computeCompleteResolve = resolve;
+			});
+		}
+		this.pending = true;
+
+		const promise = this._compute(...dependencies);
+		this._latestPromise = promise;
+
+		(async() => {
+			let newValue;
+			let success;
+			let error = null;
+
+			try {
+				newValue = await promise;
+				success = true;
+			} catch (err) {
+				error = err;
+				success = false;
+			}
+
+			// Make sure this promise is still the latest so as not to overwrite a more recent async value
+			if (this._latestPromise === promise) {
+				if (success) {
+					this.value = newValue;
+				}
+
+				this.success = success;
+				this.error = error;
+				this.pending = false;
+				this.host.requestUpdate();
```

Comment:
I'm not sure if Lit has any best practices for this... wondering if it's worth passing something to `requestUpdate` so the consumer can distinguish this call in their `updates(changedProperties)`. Just in case they want to specifically take an action when this controller triggers an update. 🤔

Reply:
Hmmm, maybe I can add an optional `id` that can be passed to `ComputedValue`'s constructor? And then maybe `ComputedValues` can use the name to pass `ComputedValues.${name}` to the id?

Reply:
Yeah I was thinking something involving the `name` as well!

Reply:
The only thing about using the name is that you can technically have more than one `ComputedValues` controller instantiated within a single component, so there could in theory be name conflicts unless we add a way to differentiate multiple `ComputedValues` controllers. That being said, I'm not 100% sure why you would want multiple. I guess in theory you could wrap it in something and reuse it, but I don't know why you'd reuse the whole `ComputedValues` controller instead of just a single `ComputedValue` one.

Reply:
One use case might be if it were used in distinct mixins that weren't aware of each other but were both applied to a component? I'm still not especially worried... if it becomes an issue then they could either be changed to use unique names or we could add an optional name/key for the entire collection, so it could become `ComputedValues.${someKey}.${name}`.

Reply:
Hmmm, I'm looking a little more closely at the `requestUpdate` [documentation](https://lit.dev/docs/api/LitElement/#LitElement.requestUpdate) and [source code](https://github.com/lit/lit/blob/a4558769b2c4008f3e15cde469280a7a1960cd3d/packages/reactive-element/src/reactive-element.ts#L1150), and I'm starting to wonder if this is a good idea. The docs state that "This should be called when an element should update based on some state not triggered by setting a reactive property. **In this case, pass no arguments**." It looks like there's a bunch of logic that's specific to Lit Properties when you pass arguments in.

We could implement our own `shouldRequestUpdate()` which can optionally be passed to `ComputedValue` to decide whether to call `requestUpdate()` (similar to `shouldCompute()` and `compute()`). It might not cover every use-case, but I image it would be good enough for most of them.

Reply:
Interesting, OK cool. Sucks that controllers have no way to indicate that they were the source of the update, but gotta go with their recommendation!

Reply:
Okay, I've added a `shouldRequestUpdate()` function that can be optionally used now.